openslides_backend/shared/env.py

```diff
@@ -1,28 +1,93 @@
-    return is_truthy(dev)
+class Environment:
+    """
+    Contains all environment variables. See the vars attribute for all defaults.
+    """
+
+    vars = {
+        "ACTION_PORT": "9002",
+        "DATASTORE_READER_HOST": "localhost",
+        "DATASTORE_READER_PATH": "/internal/datastore/reader",
+        "DATASTORE_READER_PORT": "9010",
+        "DATASTORE_READER_PROTOCOL": "http",
+        "DATASTORE_WRITER_HOST": "localhost",
+        "DATASTORE_WRITER_PATH": "/internal/datastore/writer",
+        "DATASTORE_WRITER_PORT": "9011",
+        "DATASTORE_WRITER_PROTOCOL": "http",
+        "INTERNAL_AUTH_PASSWORD_FILE": "",
+        "MEDIA_HOST": "localhost",
+        "MEDIA_PATH": "/internal/media",
+        "MEDIA_PORT": "9006",
+        "MEDIA_PROTOCOL": "http",
+        "OPENSLIDES_BACKEND_COMPONENT": "all",
+        "OPENSLIDES_BACKEND_NUM_WORKERS": "1",
+        "OPENSLIDES_BACKEND_RAISE_4XX": "false",
+        "OPENSLIDES_BACKEND_WORKER_TIMEOUT": "30",
+        "OPENSLIDES_DEVELOPMENT": "false",
+        "OPENSLIDES_LOGLEVEL": Loglevel.NOTSET.name,
+        "PRESENTER_PORT": "9003",
+        "VOTE_HOST": "vote",
+        "VOTE_PATH": "/internal/vote",
+        "VOTE_PORT": "9013",
+        "VOTE_PROTOCOL": "http",
+    }
+
+    def __init__(self, os_env: Any, *args: Any, **kwargs: Any) -> None:
+        for k, v in list(self.vars.items()):
+            env = os_env.get(k)
+            if env is not None:
+                self.vars[k] = env
+
+        # Extend the vars attribute with the lower case properties for the service URLs.
+        for service in ("datastore_reader", "datastore_writer", "media", "vote"):
+            key = service + "_url"
+            self.vars[key] = self.get_endpoint(service.upper())
+
+    def __getattr__(self, attr: str) -> str:
+        value = self.vars.get(attr)
+        if value is None:
+            raise AttributeError(f"Environment variable {attr} not found")
+        return value
+
+    def get_address(self, view: str) -> str:
+        if view == "ActionView":
+            return f"0.0.0.0:{self.ACTION_PORT}"
+        elif view == "PresenterView":
+            return f"0.0.0.0:{self.PRESENTER_PORT}"
+        raise ValueError(f"Invalid view {view}")
 
+    def get_loglevel(self) -> str:
+        lvl = self.OPENSLIDES_LOGLEVEL.upper()
+        if lvl not in Loglevel.__members__:
+            raise ValueError(f"Invalid OPENSLIDES_LOGLEVEL: {lvl}")
+        if lvl == Loglevel.NOTSET.name:
+            if is_truthy(self.OPENSLIDES_DEVELOPMENT):
+                return Loglevel.DEBUG.name
+            return Loglevel.INFO.name
+        return lvl
 
-def get_internal_auth_password() -> str:
-    if is_dev_mode():
-        return DEV_PASSWORD
-    filename = os.environ.get(INTERNAL_AUTH_PASSWORD_FILE)
-    if filename:
-        with open(filename) as file_:
-            return file_.read()
-    else:
-        raise ServerError("No internal auth password specified.")
```

Comment:
I would keep this method in the new `Environment` class instead of moving it to the view.

Reply:
I think this is part of the action view and does not belong to generic environment variable logic. And it is only used in this view. If we reuse it elsewhere we can move it to the shared module.

Reply:
I disagree, but we can leave it this way.